internal/subproviders/morpheus/datasources/serviceplan/datasource.go

```diff
@@ -0,0 +1,207 @@
+}
+
+func getServicePlanByName(
+	ctx context.Context,
+	name string,
+	provisionTypeCode string,
+	apiClient *sdk.APIClient,
+) (*sdk.GetServicePlans200ResponseServicePlan, error) {
+	pTypes, hresp, err := apiClient.ProvisioningAPI.ListProvisionTypes(ctx).Code(
+		provisionTypeCode).Execute()
+	if pTypes == nil || err != nil || hresp.StatusCode != http.StatusOK {
+		return nil, fmt.Errorf("GET failed for service plan , provision type code %s: %s",
+			provisionTypeCode, internalErrors.ErrMsg(err, hresp))
+	}
+
+	var matchingProvisionTypes []sdk.
+		GetInstanceTypeProvisioning200ResponseAllOfInstanceTypeInstanceTypeLayoutsInnerProvisionType
+	for _, pt := range pTypes.GetProvisionTypes() {
+		if ptCode, ok := pt.GetCodeOk(); ok && *ptCode == provisionTypeCode {
+			matchingProvisionTypes = append(matchingProvisionTypes, pt)
+		}
+	}
+
+	if len(matchingProvisionTypes) == 0 {
+		return nil, fmt.Errorf("provision type with code %s not found", provisionTypeCode)
+	}
+
+	if len(matchingProvisionTypes) > 1 {
+		return nil, fmt.Errorf("multiple provision types with code %s found", provisionTypeCode)
+	}
+
+	pTypeID, ok := matchingProvisionTypes[0].GetIdOk()
+	if !ok {
+		return nil, fmt.Errorf("id not found for provision type with code %s", provisionTypeCode)
+	}
+
+	ps, hresp, err := apiClient.ServicePlansAPI.ListServicePlans(ctx).Name(
+		name).ProvisionTypeId(*pTypeID).Execute()
+	if ps == nil || err != nil || hresp.StatusCode != http.StatusOK {
+		return nil, fmt.Errorf(
+			"GET failed for service_plan %s: %s", name, internalErrors.ErrMsg(err, hresp))
+	}
+
+	var matchingServicePlans []sdk.ListServicePlans200ResponseAllOfServicePlansInner
+
+	for _, sp := range ps.GetServicePlans() {
+		if pName, pNameOk := sp.GetNameOk(); pNameOk {
+			if pProvisionType, pProvisionTypeOk := sp.GetProvisionTypeOk(); pProvisionTypeOk {
+				// now check name and ProvisionType match getplanByName() params
+				if *pName == name && pProvisionType.GetCode() == provisionTypeCode {
+					matchingServicePlans = append(matchingServicePlans, sp)
+				}
+			}
+		}
+	}
+	if len(matchingServicePlans) == 1 {
+		if pID, pIDOk := matchingServicePlans[0].GetIdOk(); pIDOk {
+			// same return types as GetPlanByID
+			return getServicePlanByID(ctx, *pID, apiClient)
+		}
+
+		return nil, fmt.Errorf("service plan %s, id not found", name)
+	} else if len(matchingServicePlans) > 1 {
+		return nil, errors.New(ErrorMultipleServicePlans)
+	}
+
+	return nil, errors.New(ErrorNoServicePlanFound)
+}
+
+func getServicePlan(
+	ctx context.Context,
+	data ServicePlanModel,
+	apiClient *sdk.APIClient,
+) (*sdk.GetServicePlans200ResponseServicePlan, error) {
+	if !data.Id.IsNull() {
+		return getServicePlanByID(ctx, data.Id.ValueInt64(), apiClient)
+	} else if !data.Name.IsNull() && !data.ProvisionTypeCode.IsNull() {
+		return getServicePlanByName(
+			ctx, data.Name.ValueString(), data.ProvisionTypeCode.ValueString(), apiClient)
+	}
+
+	return nil, errors.New(ErrorNoValidSearchTerms)
```

Comment:
generally we should try to vie give a more informative error message, eg "you provided
name but both name and code are required". (With the valiation this may be less
important here, but worth noting as a general rule)